In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
test_data = pd.read_csv('/home/Machine Learning/week-8/dataset/test.csv')
test_data.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001015   Male     Yes          0      Graduate            No   
1  LP001022   Male     Yes          1      Graduate            No   
2  LP001031   Male     Yes          2      Graduate            No   
3  LP001035   Male     Yes          2      Graduate            No   
4  LP001051   Male      No          0  Not Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5720                  0       110.0             360.0   
1             3076               1500       126.0             360.0   
2             5000               1800       208.0             360.0   
3             2340               2546       100.0             360.0   
4             3276                  0        78.0             360.0   

   Credit_History Property_Area  
0             1.0         Urban  
1             1.0         Urban  
2             1.0         Urban  
3             NaN         Urban  
4             1.0         Urban

In [4]:
test_data.shape

(367, 12)

In [5]:
test_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 367 entries, 0 to 366
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Loan_ID            367 non-null    object 
 1   Gender             356 non-null    object 
 2   Married            367 non-null    object 
 3   Dependents         357 non-null    object 
 4   Education          367 non-null    object 
 5   Self_Employed      344 non-null    object 
 6   ApplicantIncome    367 non-null    int64  
 7   CoapplicantIncome  367 non-null    int64  
 8   LoanAmount         362 non-null    float64
 9   Loan_Amount_Term   361 non-null    float64
 10  Credit_History     338 non-null    float64
 11  Property_Area      367 non-null    object 
dtypes: float64(3), int64(2), object(7)
memory usage: 34.5+ KB


In [6]:
train_data = pd.read_csv('/home/Machine Learning/week-8/dataset/train.csv')
train_data.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y

In [7]:
train_data.shape

(614, 13)

In [8]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 614 entries, 0 to 613
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Loan_ID            614 non-null    object 
 1   Gender             601 non-null    object 
 2   Married            611 non-null    object 
 3   Dependents         599 non-null    object 
 4   Education          614 non-null    object 
 5   Self_Employed      582 non-null    object 
 6   ApplicantIncome    614 non-null    int64  
 7   CoapplicantIncome  614 non-null    float64
 8   LoanAmount         592 non-null    float64
 9   Loan_Amount_Term   600 non-null    float64
 10  Credit_History     564 non-null    float64
 11  Property_Area      614 non-null    object 
 12  Loan_Status        614 non-null    object 
dtypes: float64(4), int64(1), object(8)
memory usage: 62.5+ KB


In [9]:
train_data = train_data.drop(columns=['Loan_ID'])

In [10]:
object_columns = train_data.select_dtypes(include=['object']).columns
numerical_columns = train_data.select_dtypes(exclude=['object']).columns

print(object_columns)
print(numerical_columns)

Index(['Gender', 'Married', 'Dependents', 'Education', 'Self_Employed',
       'Property_Area', 'Loan_Status'],
      dtype='object')
Index(['ApplicantIncome', 'CoapplicantIncome', 'LoanAmount',
       'Loan_Amount_Term', 'Credit_History'],
      dtype='object')


In [ ]:
for index, c_col in enumerate(object_columns):
    plt.figure(index)
    sns.countplot(x=c_col, data=train_data, hue='Loan_Status')

In [13]:
train_df_encoded = pd.get_dummies(train_data, drop_first=True)
train_df_encoded.head()

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History  Gender_Male  Married_Yes  Dependents_1  Dependents_2  \
0             1.0            1            0             0             0   
1             1.0            1            1             1             0   
2             1.0            1            1             0             0   
3             1.0            1            1             0             0   
4             1.0            1            0             0             0   

   Dependents_3+  Education_Not Graduate  Self_Employed_Yes  \
0              0                       0                  0   
1              0                       0                  0   
2              0                       0                  1   
3              0                       1                  0   
4              0                       0                  0   

   Property_Area_Semiurban  Property_Area_Urban  Loan_Status_Y  
0                        0                    1              1  
1                        0                    0              0  
2                        0                    1              1  
3                        0                    1              1  
4                        0                    1              1

In [14]:
#encode categorical variables
train_data_encoded = pd.get_dummies(train_data, drop_first=True)
train_data_encoded.head()

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History  Gender_Male  Married_Yes  Dependents_1  Dependents_2  \
0             1.0            1            0             0             0   
1             1.0            1            1             1             0   
2             1.0            1            1             0             0   
3             1.0            1            1             0             0   
4             1.0            1            0             0             0   

   Dependents_3+  Education_Not Graduate  Self_Employed_Yes  \
0              0                       0                  0   
1              0                       0                  0   
2              0                       0                  1   
3              0                       1                  0   
4              0                       0                  0   

   Property_Area_Semiurban  Property_Area_Urban  Loan_Status_Y  
0                        0                    1              1  
1                        0                    0              0  
2                        0                    1              1  
3                        0                    1              1  
4                        0                    1              1

In [15]:
from sklearn.model_selection import train_test_split

X = train_data_encoded.drop(columns=['Loan_Status_Y'])
y = train_data_encoded['Loan_Status_Y']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [16]:
from sklearn.impute import SimpleImputer
imp = SimpleImputer(strategy='mean')
imp_train = imp.fit(X_train)
X_train = imp_train.transform(X_train)
X_test_imp = imp_train.transform(X_test)

# Logistic Regression

In [32]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.metrics import accuracy_score, f1_score

logreg = LogisticRegression()
logreg.fit(X_train, y_train)
y_pred = logreg.predict(X_test_imp)

print('Accuracy score: ', accuracy_score(y_test, y_pred))
print('F1 score: ', f1_score(y_test, y_pred))
print('Training score: ', logreg.score(X_train, y_train))


pd.crosstab(y_test, y_pred, rownames=['True'], colnames=['Predicted'], margins=True)


Accuracy score:  0.7804878048780488
F1 score:  0.8524590163934427
Training score:  0.814663951120163


Predicted   0    1  All
True                   
0          18   25   43
1           2   78   80
All        20  103  123

# SVM

In [43]:
from sklearn.svm import SVC

svc = SVC(kernel='linear')
svc.fit(X_train, y_train)
y_pred = svc.predict(X_test_imp)

In [44]:
print('Accuracy score: ', accuracy_score(y_test, y_pred))
print('F1 score: ', f1_score(y_test, y_pred))
print('Training score: ', svc.score(X_train, y_train))

pd.crosstab(y_test, y_pred, rownames=['True'], colnames=['Predicted'], margins=True)

Accuracy score:  0.7642276422764228
F1 score:  0.8465608465608466
Training score:  0.7881873727087576


Predicted   0    1  All
True                   
0          14   29   43
1           0   80   80
All        14  109  123

# KNN

In [72]:
from sklearn.neighbors import KNeighborsClassifier

knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)
y_pred = knn.predict(X_test_imp)

In [41]:
print(knn.score(X_train, y_train))

print('Accuracy score: ', accuracy_score(y_test, y_pred))
print('F1 score: ', f1_score(y_test, y_pred))
print('Training score: ', knn.score(X_train, y_train))

pd.crosstab(y_test, y_pred, rownames=['True'], colnames=['Predicted'], margins=True)

0.7372708757637475
Accuracy score:  0.5853658536585366
F1 score:  0.7272727272727272
Training score:  0.7372708757637475


Predicted   0    1  All
True                   
0           4   39   43
1          12   68   80
All        16  107  123

# Decision Tree

In [17]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score, f1_score, confusion_matrix

In [18]:
dt = DecisionTreeClassifier()
dt.fit(X_train, y_train)
y_pred = dt.predict(X_train)

In [19]:
print('Accuracy score: ', accuracy_score(y_train, y_pred))
print('F1 score: ', f1_score(y_train, y_pred))
print('Confusion matrix: ', confusion_matrix(y_train, y_pred))
print('Cross valdiation score mean: ', cross_val_score(dt, X_train, y_train, cv=5, scoring='f1_macro').mean())
print('Validation mean accuracy_score: ', cross_val_score(dt, X_train, y_train, cv=5, scoring='accuracy').mean())


Accuracy score:  1.0
F1 score:  1.0
Confusion matrix:  [[149   0]
 [  0 342]]
Cross valdiation score mean:  0.6138889345679155
Validation mean accuracy_score:  0.6822716965574108


In [69]:
dt = DecisionTreeClassifier(max_depth=7, min_samples_leaf=25)
dt.fit(X_train, y_train)
y_pred = dt.predict(X_test_imp)

In [70]:
print('Accuracy score: ', accuracy_score(y_test,y_pred))
print("Test F1 Score: ",f1_score(y_test,y_pred))
print("Confusion Matrix on Test Data")
pd.crosstab(y_test, y_pred, rownames=['True'], colnames=['Predicted'], margins=True)

Accuracy score:  0.7886178861788617
Test F1 Score:  0.8586956521739131
Confusion Matrix on Test Data


Predicted   0    1  All
True                   
0          18   25   43
1           1   79   80
All        19  104  123

# Random Forest Classifier

In [60]:
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(n_estimators=100, max_depth=3, min_samples_leaf=5)
rf.fit(X_train, y_train)
y_pred = rf.predict(X_train)

In [61]:
print('Accuracy score: ', accuracy_score(y_train, y_pred))
print('F1 score: ', f1_score(y_train, y_pred))
print('Confusion matrix: ', confusion_matrix(y_train, y_pred))
print('Cross valdiation score mean: ', cross_val_score(rf, X_train, y_train, cv=5, scoring='f1_macro').mean())
print('Validation mean accuracy_score: ', cross_val_score(rf, X_train, y_train, cv=5, scoring='accuracy').mean())

Accuracy score:  0.814663951120163
F1 score:  0.8807339449541285
Confusion matrix:  [[ 64  85]
 [  6 336]]
Cross valdiation score mean:  0.7305999484215479
Validation mean accuracy_score:  0.8125747268604411


In [29]:
y_pred = rf.predict(X_test_imp)

print('Accuracy score: ', accuracy_score(y_test,y_pred))
print("Test F1 Score: ", f1_score(y_test,y_pred))
print("Confusion Matrix on Test Data")
pd.crosstab(y_test, y_pred, rownames=['True'], colnames=['Predicted'], margins=True)

Accuracy score:  0.7886178861788617
Test F1 Score:  0.8586956521739131
Confusion Matrix on Test Data


Predicted   0    1  All
True                   
0          18   25   43
1           1   79   80
All        19  104  123

# Gradient Boosting Classifier

In [56]:
from sklearn.ensemble import GradientBoostingClassifier

gb = GradientBoostingClassifier(n_estimators=100, max_depth=7, min_samples_leaf=25)
gb.fit(X_train, y_train)
y_pred = gb.predict(X_train)

In [57]:
print('Accuracy score: ', accuracy_score(y_train, y_pred))
print('F1 score: ', f1_score(y_train, y_pred))
print('Confusion matrix: ', confusion_matrix(y_train, y_pred))
print('Cross valdiation score mean: ', cross_val_score(gb, X_train, y_train, cv=5, scoring='f1_macro').mean())
print('Validation mean accuracy_score: ', cross_val_score(gb, X_train, y_train, cv=5, scoring='accuracy').mean())

Accuracy score:  0.945010183299389
F1 score:  0.9614835948644793
Confusion matrix:  [[127  22]
 [  5 337]]
Cross valdiation score mean:  0.7184563047334157
Validation mean accuracy_score:  0.7819212533498248


# Grid Search

In [ ]:
from sklearn.model_selection import GridSearchCV

param_grid = {'n_estimators': [100, 200, 300, 400, 500], 'max_depth': [3, 5, 7, 9, 11], 'min_samples_leaf': [5, 10, 15, 20, 25]}
grid = GridSearchCV(gb, param_grid, cv=5, scoring='f1_macro')
grid.fit(X_train, y_train)

print(grid.best_params_)
print(grid.best_score_)
print(grid.best_estimator_)
print(grid.best_index_)
print(grid.cv_results_)

## Conclusion

The dataset incudes customer information of a bank. I implemented a classification model to predict whether a customer can take credit from the bank or not. We used 6 different classification models and compared their performance. The best model was Gradient Boosting Classifier with an accuracy of 0.86.

Firstly, I imported the dataset and checked the first 5 rows of the dataset. Then we checked the number of rows and columns of the dataset. We also checked the data types of the columns. I found that there are 367 rows and 12 columns in the dataset. The data types of the columns are int64, float64, object. I converted the data types of the columns to float64 and int 64 from object by using get_dummies() function from pandas. So our data was ready for the classification models. I split the dataset into train and test sets. I used 80% of the dataset for training and 20% of the dataset for testing. I used 6 different classification models and compared their performance. The best model was Gradient Boosting Classifier with an accuracy of 0.86. Accuracy score of the models are as follows: Logistic Regression: 0.78 SVM: 0.76 KNN: 0.73 Decision Tree: 0.78 Random Forest Classifier: 0.81 Gradient Boosting Classifier: 0.81. Then I added GridSearchCV to the Gradient Boosting Classifier and found the best parameters. The best parameters were: {'max_depth': 7, 'min_samples_leaf: 25', 'n_estimators': 100}. Then I used the best parameters to the Gradient Boosting Classifier and found the accuracy score. The accuracy score increased to 0.95. Others as follows:
Logistic Regression: from 0.78   
SVM: from 0.76 to 0.76
KNN: from 0.73 to 0.73
Decision Tree: from 0.78 to 0.78. No change.
Random Forest Classifier: from 0.81 to 0.78. So, I did not change it.
Gradient Boosting Classifier: from 0.81 to 0.95.
The best model was Gradient Boosting Classifier with an accuracy of 0.95.